token-lending/program/tests/flash_borrow_repay.rs

```diff
@@ -0,0 +1,1424 @@
+    let flash_loan_proxy_program_id = flash_loan_proxy_program_account.pubkey();
+    test.prefer_bpf(false);
+    test.add_program(
+        "flash_loan_proxy",
+        flash_loan_proxy_program_id.clone(),
+        processor!(helpers::flash_loan_proxy::process_instruction),
+    );
+
+    const FLASH_LOAN_AMOUNT: u64 = 1_000 * FRACTIONAL_TO_USDC;
+    const FEE_AMOUNT: u64 = 3_000_000;
+
+    let user_accounts_owner = Keypair::new();
+    let lending_market = add_lending_market(&mut test);
+
+    let mut reserve_config = test_reserve_config();
+    reserve_config.fees.flash_loan_fee_wad = 3_000_000_000_000_000;
+
+    let usdc_mint = add_usdc_mint(&mut test);
+    let usdc_oracle = add_usdc_oracle(&mut test);
+    let usdc_test_reserve = add_reserve(
+        &mut test,
+        &lending_market,
+        &usdc_oracle,
+        &user_accounts_owner,
+        AddReserveArgs {
+            user_liquidity_amount: FEE_AMOUNT,
+            liquidity_amount: FLASH_LOAN_AMOUNT,
+            liquidity_mint_pubkey: usdc_mint.pubkey,
+            liquidity_mint_decimals: usdc_mint.decimals,
+            config: reserve_config,
+            ..AddReserveArgs::default()
+        },
+    );
+
+    let (mut banks_client, payer, recent_blockhash) = test.start().await;
+
+    let initial_liquidity_supply =
+        get_token_balance(&mut banks_client, usdc_test_reserve.liquidity_supply_pubkey).await;
+    assert_eq!(initial_liquidity_supply, FLASH_LOAN_AMOUNT);
+
+    let usdc_reserve = usdc_test_reserve.get_state(&mut banks_client).await;
+    let initial_available_amount = usdc_reserve.liquidity.available_amount;
+    assert_eq!(initial_available_amount, FLASH_LOAN_AMOUNT);
+
+    let initial_token_balance =
+        get_token_balance(&mut banks_client, usdc_test_reserve.user_liquidity_pubkey).await;
+    assert_eq!(initial_token_balance, FEE_AMOUNT);
+
+    let mut transaction = Transaction::new_with_payer(
+        &[helpers::flash_loan_proxy::repay_proxy(
+            flash_loan_proxy_program_id,
+            FLASH_LOAN_AMOUNT,
+            usdc_test_reserve.user_liquidity_pubkey,
+            usdc_test_reserve.liquidity_supply_pubkey,
+            usdc_test_reserve.config.fee_receiver,
+            usdc_test_reserve.liquidity_host_pubkey,
+            usdc_test_reserve.pubkey,
+            spl_token_lending::id(),
+            lending_market.pubkey,
+            user_accounts_owner.pubkey(),
+        )],
+        Some(&payer.pubkey()),
+    );
+
+    transaction.sign(&[&payer, &user_accounts_owner], recent_blockhash);
+
+    // check that transaction fails
+    assert_eq!(
+        banks_client
+            .process_transaction(transaction)
+            .await
+            .unwrap_err()
+            .unwrap(),
+        TransactionError::InstructionError(
+            0,
+            InstructionError::Custom(LendingError::FlashRepayCpi as u32)
+        )
+    );
+}
+
+#[tokio::test]
+async fn end_to_end_with_flash_borrow() {
```

Comment:
I added this end-to-end test which verifies that the reserve has been force refreshed during flash borrow.